### ISSUE TITLE
requirements compiled from pyproject.toml not regenerated if separate .in file exists

### ISSUE THREAD

**Issue by danielroseman** [dependabot/dependabot-core#8365]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

pip-compile

### Package manager version

pip-compile 7.3.0

### Language version

Python 3.10

### Manifest location and content before the Dependabot update

[pyproject.toml](https://github.com/danielroseman/deptest/blob/master/pyproject.toml)
[requirements.ci.in](https://github.com/danielroseman/deptest/blob/master/requirements.ci.in)
[requirements.ci.txt](https://github.com/danielroseman/deptest/blob/master/requirements.ci.txt)
[requirements.txt](https://github.com/danielroseman/deptest/blob/master/requirements.txt)
[requirements.dev.txt](https://github.com/danielroseman/deptest/blob/master/requirements.dev.txt)

### Updated dependency

typer 0.70 -> 0.90
google-cloud-aiplatform 1.24.0 -> 1.36.0
transformers 4.27.4 -> 4.36.0

### What you expected to see, versus what you actually saw

The project uses both pyproject.toml (for production and development requirements) and a separate requirements.ci.in for use on CI. Both are compiled to requirements.* files using pip-compile.

Dependabot successfully found and updated the dependencies in both input files, pyproject.toml and requirements.ci.in. However it only regenerated the CI output file, it did not regenerate the files compiled from pyproject.toml.

Interestingly, testing it with a separate branch (pyproj) which did not include the CI files caused dependabot to successfully recreate the main and dev requirements txts - so it appears it's the addition of the separate .ci.in that confuses it.

### Smallest manifest that reproduces the issue

Linked above - this is a dummy project solely to demonstrate the problem, as the real repo is private.

**Comment by deivid-rodriguez** [comment]
Yeah, currently if we detect any `.in` files, we'll use pip-compile to update the corresponding files replacing the `.in` extension with `.txt`.

Instead, we should read the `.in` file header command, and run just that.

I suspect the case without the `.in` file worked by chance, because all files got updated directly (without using pip-compile at all).

**Comment by avilaton** [comment]
> Yeah, currently if we detect any `.in` files, we'll use pip-compile to update the corresponding files replacing the `.in` extension with `.txt`.
> 
> Instead, we should read the `.in` file header command, and run just that.
> 
> I suspect the case without the `.in` file worked by chance, because all files got updated directly (without using pip-compile at all).

This would help implementing uv since uv compiles dependencies and adds the command to the requirements files it generates.

**Comment by sirosen** [comment]
As a quick note on this, we are seeing this in `pip-tools`'s repo itself now.
We recently adjusted our doc requirements to include dependencies of the project itself (from `pyproject.toml`). Dependabot seems to run `pip-compile` with its own autodetection logic, resulting in bad update PRs, e.g.: https://github.com/jazzband/pip-tools/pull/2306

If you look at the current header of the `docs/requirements.txt` file, the `pip-compile` command is recorded in the header there as
```
#    pip-compile --allow-unsafe --output-file=docs/requirements.txt --strip-extras ./pyproject.toml docs/requirements.in
```
which would work correctly (with `-P` added, as appropriate), if it were run from the repo root.

But when I look at the Dependabot run, I see that it's [running this command](https://github.com/jazzband/pip-tools/actions/runs/20808177329/job/59766478592#step:3:293):
```
pyenv exec pip-compile -v --build-isolation --allow-unsafe --output-file\=requirements.txt -P urllib3 requirements.in {}
```

I understand not wanting to run an arbitrary command which the user supplies via that header, but perhaps Dependabot could
- parse that line
- ignore it if it's not a `pip-compile` command
- see if the output file is set to a path (gets you the correct base dir to run from)
- run a modified copy of that command
?

I'd be interested in anything we could do from the `pip-tools` side to make this work better.
